asf_heat_pump_suitability/analysis/evaluate_hp_suitability_outputs/pandera_eval_hps_outputs.py

```diff
@@ -0,0 +1,174 @@
+    add(c, float, Check.in_range(0, 1), required=True, nullable=True)
+
+# -- boolean ---------------------------------------------------------------- #
+for c in cfg.BOOLEAN_COLUMNS:
+    add(c, pl.Boolean, required=True, nullable=True)
+
+# -- categorical ------------------------------------------------------------ #
+for c, allowed in cfg.CATEGORICAL_COLUMNS.items():
+    add(c, str, Check.isin(allowed), required=True, nullable=True)
+
+# -- non-negative ----------------------------------------------------------- #
+for c in cfg.NON_NEGATIVE_COLUMNS:
+    add(c, float, Check.ge(0), required=True, nullable=True)
+
+# -- 0–100 percentage ------------------------------------------------------- #
+add(
+    "heatpump_installation_percentage",
+    float,
+    Check.in_range(0, 100),
+    required=True,
+    nullable=True,
+)
+
+# -- lsoa name -------------------------------------------------------------- #
+add("lsoa_name", str, required=True, nullable=True)
+
+# finally create the schema
+schema = DataFrameSchema(schema_columns, strict=False, coerce=True)
+
+
+# ------------------------------------------------------------------------------
+# 2. LOAD WITH POLARS
+# ------------------------------------------------------------------------------
+try:
+    df_pol = pl.read_csv(cfg.DATA_PATH)
+    logging.info("Data loaded successfully with Polars!")
+except FileNotFoundError:
+    logging.error(f"File not found at path: {cfg.DATA_PATH}")
+    raise
+except Exception:
+    logging.exception(f"Could not load data from {cfg.DATA_PATH}")
+    raise
+
+# ------------------------------------------------------------------------------
+# 3. VALIDATE WITH PANDERA (Polars backend)
+# ------------------------------------------------------------------------------
+logging.info("=== Validating with Pandera Schema ===")
+try:
+    df_valid = schema.validate(df_pol, lazy=True)
+    logging.info("Data validated successfully with Pandera!")
+except pa.errors.SchemaErrors as err:
+    logging.warning("Pandera validation failed!")
+    logging.warning(json.dumps(err.message, indent=2))
+    logging.info("Continuing with unvalidated data.")
+    df_valid = df_pol.clone()
+
+
+# ------------------------------------------------------------------------------
+# 4. OUTLIER DETECTION (Z-Score) IN POLARS
+# ------------------------------------------------------------------------------
+# compute z-score columns for numeric columns
+df_valid = df_valid.with_columns(
+    [
+        ((pl.col(c) - pl.col(c).mean()) / pl.col(c).std()).alias(f"{c}_z")
+        for c in cfg.NUMERIC_COLUMNS
+    ]
+)
+logging.info("=== Outlier Detection (Z-Score) ===")
+
+Z = cfg.OUTLIER_ZSCORE_THRESHOLD
+num_cols = cfg.NUMERIC_COLUMNS
+
+
+# outlier counts (1-row wide → long)
+outlier_counts = df_valid.select(
+    [(pl.col(f"{c}_z") > Z).sum().alias(c) for c in num_cols]
+).transpose(include_header=True, header_name="variable", column_names=["n_out"])
```

Comment:
```suggestion
).transpose(include_header=True, header_name="variable", column_names=["n_outliers"])
```
I can see myself being confused about this in the logs lol

Reply:
Good shout loool